riot/riot.py

```diff
@@ -332,49 +461,55 @@ def run(
         )
 
         for inst in self.venv.instances(pattern=pattern):
-            if pythons and inst.py not in pythons:
-                logger.debug(
-                    "Skipping venv instance %s due to interpreter mismatch", inst
-                )
-                continue
+            # Result which will be updated with the test outcome.
+            result = VenvInstanceResult(instance=inst)
 
             try:
-                base_venv_path: str = inst.py.venv_path()
-            except FileNotFoundError:
-                if skip_missing:
-                    logger.warning("Skipping missing interpreter %s", inst.py)
+                if inst.interpreter is None:
+                    if skip_missing:
+                        logger.warning(
+                            f"Skipping %s due to missing interpreter for '{inst.interpreter_hint}'",
+                            inst,
+                        )
+                        continue
+                    else:
+                        raise RunFailure(
+                            f"Missing interpreter for '{inst.interpreter_hint}'"
+                        )
+                if pythons and inst.interpreter not in pythons:
+                    logger.debug("Skipping %s due to interpreter mismatch", inst)
                     continue
-                else:
-                    raise
 
-            logger.info("Running with %s", inst.py)
+                logger.info("Running with %s", inst.interpreter)
+                base_venv_path = inst.interpreter.venv_path()
 
-            # Resolve the packages required for this instance.
-            pkgs: t.Dict[str, str] = {
-                name: version for name, version in inst.pkgs if version is not None
-            }
+                # Resolve the packages required for this instance.
+                pkgs: t.Dict[str, str] = {
+                    name: version for name, version in inst.pkgs if version is not None
+                }
```

Comment:
I found transforming/relocating this and `pkg_str` into `VenvInstance` much nicer to work with.